### ISSUE TITLE
Table (Virtual Scroll Lazy Loading) is never loading the second data page depending on scrollHeight

### ISSUE THREAD

**Issue by vtecc64** [primefaces/primeng#17183]
### Describe the bug

Depending on the scrollHeight (in comparison to the columnHeight and rows settings I guess) the second page of data gets never loaded in a Virtual Scroll Lazy Loading Table setup:

![PrimeNG Virtual Scroll Lazy Loading Table 2nd Page Problem](https://github.com/user-attachments/assets/38cd964e-a07d-4991-ba0e-86a81ba66339)

All of the following pages of data will get loaded and displayed, just the second page is missing.

### Reason for not contributing a PR

- [ ] Lack of time
- [ ] Unsure how to implement the fix/feature
- [X] Difficulty understanding the codebase
- [ ] Other

### Other Reason

I'm an Angular beginner 🤓

### Reproducer

https://6fcityg7.stackblitz.io

### Environment

Reproduceable on the StackBlitz example taken directly from the PrimeNG Table documentation page: https://primeng.org/table#virtual-scroll-lazy

### Angular version

18.0.2

### PrimeNG version

v19

### Steps to reproduce the behavior

Modify the `scrollHeight` in the StackBlitz example in `table-virtual-scroll-lazy-demo.html` to 250 or higher:

`scrollHeight="250px"`

`table-virtual-scroll-lazy-demo.html`:

```
<div class="card">
  <p-table
    [columns]="cols"
    [value]="virtualCars"
    [scrollable]="true"
    scrollHeight="250px"
    [rows]="10"
    [virtualScroll]="true"
    [lazy]="true"
    (onLazyLoad)="loadCarsLazy($event)"
  >
    <ng-template pTemplate="header" let-columns>
      <tr>
        <th *ngFor="let col of columns" style="width: 20%">{{col.header}}</th>
      </tr>
    </ng-template>
    <ng-template pTemplate="body" let-rowData let-columns="columns">
      <tr style="height: 46px">
        <td *ngFor="let col of columns">{{rowData[col.field]}}</td>
      </tr>
    </ng-template>
    <ng-template pTemplate="loadingbody" let-columns="columns">
      <tr style="height: 46px">
        <td *ngFor="let col of columns; let even = even">
          <p-skeleton
            [ngStyle]="{'width': even ? (col.field === 'year' ? '30%' : '40%') : '60%'}"
          />
        </td>
      </tr>
    </ng-template>
  </p-table>
</div>
```

### Expected behavior

Adjusting the scroll height to 250 and above shall not prevent the second data page from being loaded and displayed. With variable row heights the problem might get even more unpredicatable.

**Comment by qwadrox** [comment]
@vtecc64 There is something with the reproducer URL. It's not opening.

**Comment by juri-sinitson** [comment]
@vtecc64 

According to the [docs](https://primeng.org/table#virtual-scroll-lazy):
> It is also suggested to use the same virtualScrollItemSize value on the tr element inside the body template.

If you add the prop `virtualScrollItemSize` you'll get that issue solved (I've reproduced your problem by omitting that prop). 
Is this issue solved for you @vtecc64?

**Comment by vtecc64** [comment]
@juri-sinitson

> You don't wanna have the constraint of the item size especially in two places: in the table template and in the table prop virtualScrollItemSize. Instead you'd like to have a reasonable computed default value for virtualScrollItemSize, right?

Exactly. In many PrimeNG examples, tables include images, which can cause rows to vary in height and take up more space than anticipated. Additionally, when loading column content dynamically—such as through native federation—it becomes challenging to predict the content and, consequently, the row heights. Therefore, having a computed default value for virtualScrollItemSize that adapts based on the actual content would be ideal. This approach would eliminate the need to manually set item sizes in both the table template and the virtualScrollItemSize property, ensuring better flexibility and compatibility across different scenarios.

I really appreciate your help!

**Comment by vtecc64** [comment]
**Additional Information for easy reproduction based on the official PrimeNG example:**

In order to be able to easily reproduce the issue lined out above I want to share my approach here. This might take 3-4 minutes or so (less if the official PrimeNG examples on StackBlitz would get fixed):

1. open the official PrimeNG StackBlitz example for the Lazy Loading Virtual Scroll Table here: https://primeng.org/table#virtual-scroll-lazy
2. It seems that many of the current StackBlitz examples are broken so we have to tweak them a little bit to make it work again:

fix the broken import in `table-virtual-scroll-lazy-demo.ts`:

`import { CarService } from '@/service/carservice';` -> `import { CarService } from '../service/carservice';`

remove the invalid import (and 2 references) regarding `'DynamicDialog'` from `imports.ts` (don't know what the replacement for it would be):

`import { DynamicDialog } from 'primeng/dynamicdialog';`

and fix the carservice import here too.


Now back to the actual issue:

3. Lets say I dont want to have 100 entries lazy loaded because they take some time to load and I decide to load 10 items. This is the only change I make:

`[rows]="10"`

And bum - the second page is not lazy loaded any more and cannot be displayed. Only items 1-10 and 21ff will be displayed.

@qwadrox maybe deriving an example from the official StackBlitz example helps more than a tailored StackBlitz example

**Comment by juri-sinitson** [comment]
As far as I see [here](https://github.com/primefaces/primeng/blob/df745663da783e8b499db2240e843ed5a69fe899/packages/primeng/src/scroller/scroller.ts#L958), the current approach to find out which part of the data (the start and the end index) should be shown on scrolling is:
1. Find out how much items fit to the visible area of the table container including the tolerance
2. Map the current scroller position to the appropriate part of the data
3. Show/load this that appropriate data part

This approach needs fixed row height of course when we're talking about vertical scrolling.

I'm thinking about implementing an extended approach for unknown height of the rows with data loaded, which can be used when the appropriate flag is set:
1. Reuse the approach above for the dummy rows (the ones which depict data loading)
2. When the data was loaded, check how much dummy rows are in the visible area
3. Load the missing data

Note: in my approach we may still need the fixed height for the dummy rows and maybe the fixed minimum height for the rows with data. If that approach works for vertical scrolling, it should actually work for the horizontal one too.

I currently blindly estimate 6 weeks for this issue with my current capacity.
I think the 2. and 3. in my approach may take time and may even be challenging.
Also automated testing of scrolling might be challenging too.
I hope it becomes better estimable when I start implementing and figure out how to test scrolling automatically.

As the next step I'm gonna check out Angular Material if they implemented the scenario with rows with data of unknown row height.

Feel free to share you thoughts any time, I'll consider them.

**Comment by juri-sinitson** [comment]
# Update
[This](https://virtuoso.dev/) library (see also [this](https://virtuoso.dev/hello/#list-with-totalcount) example) looks promising to inspire for the implementation of virtual+lazy loading on scroll without fixed height of component depicting data. As next step I'm analyzing how they're achieving  this.

**Comment by juri-sinitson** [comment]
# UPDATE
Due to different spontaneous private issues I have to postpone.
As soon as I have progress on this issue, I'll give a short feedback.

**Comment by marcin-bury** [comment]
@juri-sinitson 
I'm new to PrimeNG and Angular itself. I've been playing with p-table lazy loading and virtual scrolling last couple of days. I wonder why the example with cars (from documentation) works but one with real data retrieving from the backend does not. In my case actually the onLazyLoad event is triggered only once. Do you know under what conditions it is triggered? 
Thanks i advance...

**Comment by juri-sinitson** [comment]
@marcin-bury 
Please provide me your code example in stackblitz or at least post the template here, that I can see the input you provide to the table. Be sure to use the same version (the last stable one) of PrimeNG they use in the example. I will check out when the `onLazyLoad` the next week and write back.

**Comment by marcin-bury** [comment]
@juri-sinitson 
Unfortunatelly I have failed setting up the project on stackblitz, however here is the link to public project on github (if you don't mind): [test-primeng-table-lazy-load](https://github.com/marcin-bury/test-primeng-table-lazy-load)
I have also added the "cars example", so we could see PrimeNG example actually works in latest version.
Thanks in advance for looking into it.

**Comment by juri-sinitson** [comment]
@marcin-bury 
According to [this](https://primeng.org/virtualscroller#basic) you need to initialize the array mapped to the table with the maximum length. Otherwise the end user won't be able to produce a scroll event. See your class below with my modifications. Other modifications of mine are rather small ones.

I faced these current limitations of PrimeNG when modifying your class:
1. When you scroll to the last table record there is kinda jitter
2. It looks like one can't currently remove the pages invisible. It seems to work well when scrolling to the bottom. But when scrolling to the top it triggers the scroll events too late and the scroll event for the first page is not triggered at all. The latter causes an empty table.

Feel free to open issues on those limitations (if there are no issues opened yet) and connect them with this one.

And here is the class. Hope it helps.

```typescript
export class DummyJsonComponent {
  totalRecords: number = 200;
  products: Product[] = Array.from({ length: this.totalRecords });
  loading: boolean = false;

  virtualRowHeight = 32;
  virtualRows = 100;

  aService = inject(DummyJsonService);

  loadProducts($event: TableLazyLoadEvent) {
    console.log('Event:', $event);

    const rows = $event.rows ?? 0;

    this.loading = true;
    this.aService
      .getProducts($event.first || 0, rows)
      .subscribe((response: ProductsResponse) => {        
        console.log('Products: ', response.products);
        
        const firstIndexOfPage = $event.first ?? 0;
        
        // When removing the unvisible rows, the following happends:
        // Works well, when scrolled to bottom, but bugs when scrolled to top.
        // In this case the scroll event is triggered too late and is 
        // not triggered for the first page.
        this.products = [
          ...this.products.slice(0, firstIndexOfPage),          
          // See the comment above
          // ...Array.from({ length: firstIndexOfPage }) as Product[],          
          ...response.products,                    
          ...this.products.slice(firstIndexOfPage + response.products.length),          
          // See the comment above
          // ...Array.from({ length: this.totalRecords - firstIndexOfPage - response.products.length }) as Product[],
        ];

        console.log('Products page: ', this.products);

        this.loading = false;
        if ($event.forceUpdate) {
          $event.forceUpdate();
        }
      });    
  }
}
```
